### ISSUE TITLE
Rejected promise on createOrder results in unhandled error

### ISSUE THREAD

**Issue by nickbrowne** [paypal/paypal-checkout-components#1095]
### Description

We have an issue where the PayPal Smart Button internals don't gracefully handle rejected promises when creating an order, resulting in spurious error reports to Bugsnag. 

Turning off or limiting what bugsnag reports is not really an option because we obviously do want to be alerted immediately if there is a legitimate frontend issue with payments. Ideally the PayPal code would not raise spurious errors - and handle rejected promises more gracefully. It seems like it shouldn't assume that all promises will simply be resolved.

Now it could just be that we're not doing something right, because it's very difficult to understand the documentation. There is (seemingly) no central API reference for smart buttons, and the few examples that do exist seem incomplete and/or overly optimistic that you won't need to handle errors. React documentation in particular is very lacking.

Handling errors as described here, which both works and is ignored at the same time:
https://developer.paypal.com/docs/checkout/integration-features/handle-errors/

This gives a very basic example of server integration but no mention of error handling:
https://developer.paypal.com/demo/checkout/#/pattern/server

Same here:
https://developer.paypal.com/docs/checkout/reference/server-integration/set-up-transaction/#on-the-client

### Issue demonstration

https://jsfiddle.net/szxeon3f/ 

 - Open console
 - Click button
 - See unhandled errors

### Paypal version
```
"5.0.29"
```

### Tested with:
Firefox 66.0.3
Chrome Version 70.0.3538.77

The the issue does not occur at [developer.paypal.com/demo/checkout](https://developer.paypal.com/demo/checkout) because the demo assumes that the promise will always be resolved.

**Comment by bluepnume** [comment]
If the user clicks on the button, and an order id can not be created for any reason -- that is by definition an error case, since we're unable to open the popup and show the checkout experience.

Can you help me understand when/why you are rejecting the promise returned to `createOrder`? There may be a better way to achieve what you're trying to do.

**Comment by nickbrowne** [comment]
We're creating a Payment ([either one off or recurring, depending what the user selects](https://donate.theconversation.com/au)) server side in Ruby, which can fail:
```rb
# https://github.com/paypal/PayPal-Ruby-SDK#create-payment

# Build Payment object
@payment = Payment.new(payment_params)

if @payment.create
  @payment.id     # Payment Id
else
  @payment.error  # Error Hash
end
```

So making assumptions about what to do in this case (because the documentation for PayPal Smart Buttons seems to assume only a success state) we saw that:
 - Rather than just return a value in the `createOrder` callback, you can return a promise, that's good!
 - In the case where the PayPal REST API returns an error (in our case, this happens on our server, which is propagated back to the client), reject that promise so that the popup closes, since that is a fairly common pattern.

We were just surprised to see that rejecting the promise just results in an unhandled error.

I totally agree with what you're saying here:
> If the user clicks on the button, and an order id can not be created for any reason -- that is by definition an error case, since we're unable to open the popup and show the checkout experience.

I think the real issue we're having is that there appears to be no graceful way to bail out of the popup once you've started the `createOrder` sequence, seeing the unhandled error is giving us doubt that we're doing the right thing. But there's no direction in the documentation as to what the right thing is.

If this is how it's supposed to work then we can deal with that. I'm just trying to understand how those callbacks were intended to be used, because right now we're not sure.

Thanks :smiley:

**Comment by boy1dr** [comment]
Any chance a resolution was found to this problem?
I too cannot seem to prevent the createOrder/fetch promise from collapsing in to an unhandled_error.

**Comment by bluepnume** [comment]
Hi -- please let me know if using `onError` to handle this case doesn't work for a particular use case. Thanks!
